File: example/python/audio_example.py
#!/usr/bin/env python3

import logging
import signal
import sys
import time
from typing import Optional

import magicdog_y1_python as magicbot

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)

robot: Optional[magicbot.MagicRobot] = None
running = True


def signal_handler(signum, frame):
    global running
    logging.info("Received signal %s, exiting", signum)
    running = False


def print_help():
    logging.info("Audio example")
    logging.info("  1 get volume")
    logging.info("  2 set volume [0-100]")
    logging.info("  3 play TTS [text]")
    logging.info("  4 stop TTS")
    logging.info("  5 open audio stream")
    logging.info("  6 close audio stream")
    logging.info("  7 subscribe audio stream")
    logging.info("  8 unsubscribe audio stream")
    logging.info("  ? help")
    logging.info("  ESC exit")


def controller():
    return robot.get_audio_controller()


def get_volume():
    status, volume = controller().get_volume()
    if status.code != magicbot.ErrorCode.OK:
        logging.error("get_volume failed: code=%s message=%s", status.code, status.message)
        return
    logging.info("Volume: %s", volume)


def set_volume(value):
    status = controller().set_volume(int(value))
    if status.code != magicbot.ErrorCode.OK:
        logging.error("set_volume failed: code=%s message=%s", status.code, status.message)


def play_tts(text):
    tts = magicbot.TtsCommand()
    tts.id = "100000000001"
    tts.content = text
    tts.priority = magicbot.TtsPriority.HIGH
    tts.mode = magicbot.TtsMode.CLEARTOP
    status = controller().play(tts)
    if status.code != magicbot.ErrorCode.OK:
        logging.error("play TTS failed: code=%s message=%s", status.code, status.message)


def stop_tts():
    status = controller().stop()
    if status.code != magicbot.ErrorCode.OK:
        logging.error("stop TTS failed: code=%s message=%s", status.code, status.message)


def open_audio_stream():
    status = controller().open_audio_stream()
    if status.code != magicbot.ErrorCode.OK:
        logging.error("open_audio_stream failed: code=%s message=%s", status.code, status.message)


def close_audio_stream():
    status = controller().close_audio_stream()
    if status.code != magicbot.ErrorCode.OK:
        logging.error("close_audio_stream failed: code=%s message=%s", status.code, status.message)


def subscribe_audio_stream():
    origin_counter = 0
    bf_counter = 0

    def origin_callback(audio_stream):
        nonlocal origin_counter
        if origin_counter % 30 == 0:
            logging.info("Origin audio frame size=%d", audio_stream.data_length)
        origin_counter += 1

    def bf_callback(audio_stream):
        nonlocal bf_counter
        if bf_counter % 30 == 0:
            logging.info("BF audio frame size=%d", audio_stream.data_length)
        bf_counter += 1

    controller().subscribe_origin_audio_stream(origin_callback)
    controller().subscribe_bf_audio_stream(bf_callback)


def unsubscribe_audio_stream():
    controller().unsubscribe_bf_audio_stream()
    controller().unsubscribe_origin_audio_stream()


def main():
    global robot
    signal.signal(signal.SIGINT, signal_handler)

    logging.info("Robot model: %s", magicbot.get_robot_model())
    robot = magicbot.MagicRobot()

    try:
        if not robot.initialize("192.168.54.111"):
            logging.error("Failed to initialize robot SDK")
            return -1

        status = robot.connect()
        if status.code != magicbot.ErrorCode.OK:
            logging.error("Failed to connect: code=%s message=%s", status.code, status.message)
            return -1

        if not controller().initialize():
            logging.error("Failed to initialize audio controller")
            return -1

        print_help()
        while running:
            line = input("Enter command: ").strip()
            if line in ("\x1b", "esc"):
                break
            if not line:
                time.sleep(0.01)
                continue
            parts = line.split(maxsplit=1)
            key = parts[0]
            arg = parts[1] if len(parts) > 1 else ""
            if key == "1":
                get_volume()
            elif key == "2":
                set_volume(arg or 50)
            elif key == "3":
                play_tts(arg or "hello")
            elif key == "4":
                stop_tts()
            elif key == "5":
                open_audio_stream()
            elif key == "6":
                close_audio_stream()
            elif key == "7":
                subscribe_audio_stream()
            elif key == "8":
                unsubscribe_audio_stream()
            elif key == "?":
                print_help()
            else:
                logging.warning("Unknown command: %s", key)

    except (EOFError, KeyboardInterrupt):
        pass
    except Exception as exc:
        logging.error("Example failed: %s", exc)
        return -1
    finally:
        if robot:
            try:
                robot.get_audio_controller().shutdown()
                robot.disconnect()
                robot.shutdown()
            except Exception as exc:
                logging.error("Cleanup failed: %s", exc)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: example/cpp/audio_example/audio_example.cpp
#include "magic_robot.h"
#include "magic_sdk_version.h"

#include <termios.h>
#include <unistd.h>
#include <csignal>

#include <iostream>

using namespace magic::y1;

magic::y1::MagicRobot robot;

void signalHandler(int signum) {
  std::cout << "Interrupt signal (" << signum << ") received.\n";

  robot.Shutdown();
  // Exit process
  exit(signum);
}

void print_help() {
  std::cout << "Key Function Description:\n";
  std::cout << "  Audio Functions:\n";
  std::cout << "  1        Function 1: Get volume\n";
  std::cout << "  2        Function 2: Set volume\n";
  std::cout << "  3        Function 3: Play TTS\n";
  std::cout << "  4        Function 4: Stop playback\n";
  std::cout << "  Audio stream Functions:\n";
  std::cout << "  5        Function 5: Open audio stream\n";
  std::cout << "  6        Function 6: Close audio stream\n";
  std::cout << "  7        Function 7: Subscribe to audio stream\n";
  std::cout << "  8        Function 8: Unsubscribe to audio stream\n";
  std::cout << "  Wakeup Status Functions:\n";
  std::cout << "  q        Function q: Open wakeup status stream\n";
  std::cout << "  w        Function w: Close wakeup status stream\n";
  std::cout << "  e        Function e: Subscribe to wakeup status\n";
  std::cout << "  r        Function r: Unsubscribe to wakeup status\n";
  std::cout << "\n";
  std::cout << "  ?        Function ?: Print help\n";
  std::cout << "  ESC      Exit program\n";
}

int getch() {
  struct termios oldt, newt;
  int ch;
  tcgetattr(STDIN_FILENO, &oldt);  // Get current terminal settings
  newt = oldt;
  newt.c_lflag &= ~(ICANON | ECHO);  // Disable buffering and echo
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  ch = getchar();                           // Read key press
  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);  // Restore settings
  return ch;
}

void GetVolume() {
  // Get audio controller
  auto& controller = robot.GetAudioController();

  // Get volume
  int get_volume = 0;
  auto status = controller.GetVolume(get_volume);
  if (status.code != ErrorCode::OK) {
    std::cerr << "get volume failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "get volume success, volume: " << std::to_string(get_volume) << std::endl;
}

void SetVolume() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Set volume
  auto status = controller.SetVolume(50);
  if (status.code != ErrorCode::OK) {
    std::cerr << "set volume failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "set volume success" << std::endl;
}

void PlayTts() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Play speech
  TtsCommand tts;
  tts.id = "100000000001";
  tts.content = "How's the weather today!";
  tts.priority = TtsPriority::HIGH;
  tts.mode = TtsMode::CLEARTOP;
  auto status = controller.Play(tts);
  if (status.code != ErrorCode::OK) {
    std::cerr << "play tts failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "play tts success" << std::endl;
}

void StopTts() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Stop speech playback
  auto status = controller.Stop();
  if (status.code != ErrorCode::OK) {
    std::cerr << "stop tts failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "stop tts success" << std::endl;
}

void OpenAudioStream() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Open audio stream
  auto status = controller.OpenAudioStream();
  if (status.code != ErrorCode::OK) {
    std::cerr << "open audio stream failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "open audio stream success" << std::endl;
}

void CloseAudioStream() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Close audio stream
  auto status = controller.CloseAudioStream();
  if (status.code != ErrorCode::OK) {
    std::cerr << "close audio stream failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "close audio stream success" << std::endl;
}

void SubscribeAudioStream() {
  // Get audio controller
  auto& controller = robot.GetAudioController();

  // Subscribe to audio streams
  controller.SubscribeOriginAudioStream([](const std::shared_ptr<AudioStream> data) {
    static int32_t counter = 0;
    if (counter++ % 30 == 0) {
      std::cout << "Received origin audio stream data, size: " << data->data_length << std::endl;
    }
  });

  controller.SubscribeBfAudioStream([](const std::shared_ptr<AudioStream> data) {
    static int32_t counter = 0;
    if (counter++ % 30 == 0) {
      std::cout << "Received bf audio stream data, size: " << data->data_length << std::endl;
    }
  });
  std::cout << "Subscribed to audio streams" << std::endl;
}

void UnsubscribeAudioStream() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Unsubscribe from audio streams
  controller.UnsubscribeOriginAudioStream();
  controller.UnsubscribeBfAudioStream();
  std::cout << "Unsubscribed from audio streams" << std::endl;
}

void OpenWakeupStatusStream() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Open wakeup status stream
  auto status = controller.OpenWakeupStatusStream();
}

void CloseWakeupStatusStream() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Close wakeup status stream
  auto status = controller.CloseWakeupStatusStream();
}

void SubscribeWakeupStatus() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Subscribe to wakeup status
  controller.SubscribeWakeupStatus([](const std::shared_ptr<WakeupStatus> data) {
    static int32_t counter = 0;
    if (data->is_wakeup) {
      if (data->enable_wakeup_orientation) {
        std::cout << "Received wakeup status data, is_wakeup: " << data->is_wakeup << ", enable_wakeup_orientation: " << data->enable_wakeup_orientation << ", wakeup_orientation: " << data->wakeup_orientation << std::endl;
      } else {
        std::cout << "Received wakeup status data, is_wakeup: " << data->is_wakeup << std::endl;
      }
    } else {
      std::cout << "Received wakeup status data, is_wakeup: " << data->is_wakeup << std::endl;
    }
  });
}

void UnsubscribeWakeupStatus() {
  // Get audio controller
  auto& controller = robot.GetAudioController();
  // Unsubscribe from wakeup status
  controller.UnsubscribeWakeupStatus();
}

int main(int argc, char* argv[]) {
  // Bind SIGINT (Ctrl+C)
  signal(SIGINT, signalHandler);

  std::cout << "SDK Version: " << SDK_VERSION_STRING << std::endl;

  print_help();

  std::string local_ip = "192.168.54.111";
  // Configure local IP address for direct network connection and initialize SDK
  if (!robot.Initialize(local_ip)) {
    std::cerr << "robot sdk initialize failed." << std::endl;
    robot.Shutdown();
    return -1;
  }

  // Connect to robot
  auto status = robot.Connect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "connect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }
  std::cout << "Press any key to continue (ESC to exit)..."
            << std::endl;

  // Wait for user input
  while (1) {
    int key = getch();
    if (key == 27)
      break;  // ESC key ASCII code is 27

    std::cout << "Key ASCII: " << key << ", Character: " << static_cast<char>(key) << std::endl;
    switch (key) {
      // 1. Audio Functions
      case '1': {
        // Get volume
        GetVolume();
        break;
      }
      case '2': {
        // Set volume
        SetVolume();
        break;
      }
      case '3': {
        PlayTts();
        break;
      }
      case '4': {
        StopTts();
        break;
      }
      // 2. Audio stream Functions
      case '5': {
        OpenAudioStream();
        break;
      }
      case '6': {
        CloseAudioStream();
        break;
      }
      case '7': {
        SubscribeAudioStream();
        break;
      }
      case '8': {
        UnsubscribeAudioStream();
        break;
      }
      // 3. Wakeup Status Functions
      case 'Q':
      case 'q': {
        OpenWakeupStatusStream();
        break;
      }
      case 'W':
      case 'w': {
        CloseWakeupStatusStream();
        break;
      }
      case 'E':
      case 'e': {
        SubscribeWakeupStatus();
        break;
      }
      case 'R':
      case 'r': {
        UnsubscribeWakeupStatus();
        break;
      }
      case '?': {
        print_help();
        break;
      }
      default:
        std::cout << "Unknown key: " << key << std::endl;
        break;
    }
    usleep(10000);
  }

  // Disconnect from robot
  status = robot.Disconnect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "disconnect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  robot.Shutdown();

  return 0;
}